Upgrade SDK to .NET 9 / MAUI with ready support for Parse Live Queries Add Live Queries (Now Ready Separately As Well)

YBTopaz8 opened the pull request:
### **Introduction**  

**Hello Parse Community for .NET,**  

I am thrilled to submit this PR that helps modernize the **Parse SDK for .NET** by:  
1. **Upgrading it to .NET 9 and MAUI**, making it compatible with the latest features and patterns.  
2. **Introducing Live Queries** support for real-time functionality, fully rebuilt for .NET 9 and integrated with **.NET MAUI** for cross-platform development.  

This contribution is a response to one of the **most requested features** in Parse for .NET: Live Queries. By bridging this gap, I believe Parse can now fully serve the modern .NET ecosystem and attract new developers who seek real-time functionality with minimal complexity and even shine some more light on the Goodies that .NET MAUI offer with this!  

---
## Acknowledging Prior Work

Before diving into the specifics, I want to recognize the existing PR [feat: Fully support .NET Standard on various platforms.](https://github.com/parse-community/Parse-SDK-dotNET/pull/372)

While this PR doesn’t mention .NET MAUI or Live Queries, it lays the groundwork for compatibility across multiple platforms. I see it as a valuable stepping stone toward modernizing the Parse SDK. My work builds on the same ethos of broadening Parse’s reach while addressing critical developer needs like real-time updates and cross-platform stability.

---

### **Why did I finally do this PR?**  

#### **1. Live Queries for .NET - Finally!**  
Live Queries have been a critical feature in Parse’s ecosystem, enabling real-time updates to clients. However, .NET developers have been left out of this functionality due to outdated implementations and a lack of modern support (Understandably so; It's no small feat!).  

This PR:   
- Completely open doors for Live Queries support on ANY .NET platforms in Rebuilds Live Queries for .NET 9

#### **2. Parse SDK for .NET 9 and .NET MAUI**  
.NET MAUI is the future of cross-platform development as per Microsoft, allowing developers to target Android, iOS, Windows, and macOS with a single codebase. Yet, Parse lacked compatibility with this framework—until now at least as I can happily report it working on MAUI devices as low as API 30 (Will test further if need be).  

This PR:  
- Upgrades the Parse SDK to **.NET 9**, modernizing its async patterns, error handling, and overall architecture.  
- Resolves critical crashes related to **`HostManifestData`** that previously made MAUI integration unstable.  
- Hopes to maintain Parse as a first-class citizen in the .NET ecosystem, empowering developers to build cross-platform and now realtime apps with Parse as their backend.
- Did some (I think most TODOs) in the codes and tested modern approaches (if the previously commented Idea was valid for its time but no longer now, I simply did a "Modern" implementation)

I had this ready for a while but wasn't done testing in my "ways" (you will understand just below) so now that I'm very happy and confident of the outcome, I am happily sharing.



### **(Somewhat) Technical Contributions (The codes might speak better than my words)**  

This PR makes several key improvements to the SDK and Live Queries:  

#### **SDK Enhancements**  
1. **Async Modernization:**  
   - Replaced legacy async patterns (`.Unwrap()`, `.OnSuccess()`) with `async/await`, improving readability, performance, and maintainability.  
   - Updated control flows to align with modern .NET best practices.  

2. **Error Handling Improvements:**  
   - Errors now return detailed error objects instead of throwing vague exceptions. This improves debugging and provides actionable insights during development.  
   - Handled several internal bugs, like the one responsible for getting `response.length()` ALWAYS throwing an exception because the method isn't supported.

3. **Minimal Dependencies:**  
   - No additional libraries were added. The SDK remains lightweight and efficient.  

4. **MAUI Compatibility:**  
   - Resolved crashes by manually setting `HostManifestData` during initialization.  
   - Fully tested on Windows, Android, and macOS (iOS untested by me).  
   - SDK should AS Secured if not theorically more secured than currently as it simply now has the modern methods (and very minimal needed changes)

#### **As my Live Queries repo is also ready if needed, I'll just name the Live Queries Enhancements I did as well**  
1. **WebSocket Overhaul:**  
   - Replaced the outdated WebSocket implementation with `System.Net.WebSockets.Client`, and all the related methods, ensuring stability and scalability.  
   - To avoid nerding out as the codes would do best for me, I simply updated the project to follow the new .NET flows/standards as much as I could

2. **Real-Time Reliability:**  
   - Thorough testing ensured stability during connection loss, reconnections, and high-frequency updates.  

3. **Feature Parity:**  
   - Integrated feature parity with other Live Query implementations in Parse’s ecosystem while aligning with .NET’s architecture.  

---

### **Challenges and My Approach**  

#### **1. Addressing Legacy Code:**  
Porting Live Queries from .NET 2 involved unraveling contextually outdated async patterns, undocumented flows, and unsupported APIs. I approached this by:  
- Just being the debugger's best buddy ! (jk! well..)
- I simply ensured to to use breakpoints on each step and the call stack to understand how the commands flow and I used VS's tool to visualize thread interactions (when testing for concurrency and most critical, low level bugs/vulnerabilities that were done)
- Refactoring methods to fit .NET 9’s architecture while preserving the core functionality.  

#### **2. Ensuring Cross-Platform Stability:**  
MAUI introduces unique challenges, particularly with platform-specific configurations like `HostManifestData`. I resolved this by implementing a manual initialization process that ensures stability across Windows, Android, and macOS.  

#### **3. Testing and Documentation:**  
While this PR lacks comprehensive unit tests, I’ve included:  
- A sample chat application demonstrating Live Queries in action (built with .NET MAUI).  
- Documentation for setting up Live Queries and Parse SDK in .NET 9/Maui projects.  

I welcome collaboration to expand test coverage and improve documentation further.  

---

## Why No Unit Tests?
I want to address this head-on because I know it’s an important topic.

The truth is:

I’ve never had the opportunity to learn unit testing properly. It’s something I’ve always envied in others who do it effortlessly.
My coding experience has been shaped by a more hands-on, old-school approach—debugging and breaking my code manually to understand and refine it.
While this approach has worked well for me so far, I recognize the value of formalized testing. Most methods in the updated SDK retain the same flow, so I believe existing test suites (if available) would require minimal adjustment. I welcome collaboration from the community to help add meaningful tests.
- Platforms I tested on:  
  - **Windows, Android, macOS** (stable).  
  - **iOS** (untested but expected to function consistently).  
- Included sample app: A simple chat application demonstrating Live Queries integration in .NET MAUI.  


## On Adherence to Guidelines:
While I didn’t strictly follow all project guidelines, I maintained the core functionality and flow of the SDK. My focus was on modernizing outdated patterns and ensuring stability, especially for .NET MAUI compatibility and Live Queries.

Any deviations (e.g., updated async patterns or additional comments) were intentional, aimed at improving readability, maintainability, and scalability. I look forward to working with the community to align this further with Parse’s standards.

---

### **Acknowledgments**  

I want to thank:  
- **The Parse Team and Parse DOTNET SDK Team:** Your work inspired me to take on this challenge and helped me get out of a situation by helping me put food on the table, and I’m honored to contribute and help others do same!
- **Back4App:** Thank you for your support and great platform (All I will say, is Check it out!)
- **[JonMcPherson](https://github.com/JonMcPherson):** Your foundational work on Live Queries laid the groundwork for this project.

---

### **Next Steps**  

This PR is just the beginning. I’m eager to collaborate with the Parse community to:  
1. Expand unit testing for both the SDK and Live Queries.  
2. Optimize the architecture further to align with Parse’s standards.  
3. Explore additional features and enhancements that benefit the broader developer base (As I personally use it, I have a personal commitment to make Parse better for my uses so my contribution is one that cna be counted on)
---

### **Final Thoughts**  

This PR isn’t about pushing a solution; it’s about opening a door. Live Queries, .NET 9 AND MAUI support represent the next step forward for Parse, and I’m excited to help make it happen.  

Hopefully at least it helps ANYONE in need and closes that near-decade old issue 

**Happy coding,**  

Yvan Brunel  

---

### **Links**  
- [Requested Feature (GitHub Issue)](https://github.com/parse-community/Parse-SDK-dotNET/issues/199)
- [Parse SDK for .NET 9 and MAUI (GitHub Repo)](https://github.com/YBTopaz8/Parse-SDK-dotNET)  
- [MY Live Queries for Parse (GitHub Repo)](https://github.com/YBTopaz8/parse-live-query-dotnet)  
- [Sample Chat App for MAUI on Windows and Android with Live Queries (Having a YT link showing how to use LQ in .NET MAUI)](lhttps://github.com/YBTopaz8/Chat-App-.NET-MAUI-LIVE-QUERY)

mtrezza commented:
Amazing, a lot of time must have gone into this. It will take some time to review this, but I've at least be able to start the CI to see how that goes.

mtrezza reviewed (commented):
Could you take a look at the failing CI?

YBTopaz8 commented:
Hi !
Thanks.

From what I see, the build fails mainly due to the UnitTest likely looking to validate the previous methods (I can see it mention `unwrap()` )
Fortunately for _me_ , I started learning on UT so I should be able to at least update those tests so that they pass for a successful merge.

Please give me some time and I should be committing soon.

mtrezza commented:
@YBTopaz8 Would you like me to add you to the Parse .NET SDK review team? You'd get a notification when a contributor's PR is ready for review.

mtrezza commented:
Great, restarted the CI...

dblythy commented on Parse/Infrastructure/CacheController.cs:
```diff
@@ -10,253 +10,304 @@
+            {
+                Storage[key] = value;
+            }
+            await SaveAsync();
```
Previous implementation blocks concurrent threads from modifying Storage and initiating SaveAsync() until SaveAsync() returns the Task. This can cause potential deadlocks if SaveAsync() depends on the same lock.

New implementation only blocks concurrent threads while updating Storage. SaveAsync() is executed outside the lock, avoiding deadlocks but risking race conditions if SaveAsync() accesses Storage.

Is this behavior change intended?

Would suggest maybe implementing a semaphore for better asynchronous synchronization.

YBTopaz8 replied:
On this now

dblythy commented on Parse/Infrastructure/CacheController.cs:
```diff
@@ -10,253 +10,304 @@
-                {
-                    operation.Invoke();
-                }
+                // Ensure operation runs within a locked context
```
This does nothing right?

YBTopaz8 replied:
I was supposed to ensure operations go on well, but that class's porting wasn't fully done from what I see, will readjust with more locks.
Just a moment (to test too)

dblythy commented on Parse/Infrastructure/CacheController.cs:
```diff
@@ -10,253 +10,304 @@
+    /// <returns>A task that completes once the file move operation form <paramref name="originFilePath"/> to <paramref name="targetFilePath"/> completes.</returns>
+    public async Task TransferAsync(string originFilePath, string targetFilePath)
+    {
+        if (!String.IsNullOrWhiteSpace(originFilePath) && !String.IsNullOrWhiteSpace(targetFilePath) && new FileInfo(originFilePath) is { Exists: true } originFile && new FileInfo(targetFilePath) is { } targetFile)
```
```suggestion
        if (!string.IsNullOrWhiteSpace(originFilePath) && !string.IsNullOrWhiteSpace(targetFilePath) && new FileInfo(originFilePath) is { Exists: true } originFile && new FileInfo(targetFilePath) is { } targetFile)
```

dblythy commented on Parse/Infrastructure/Control/ParseAddOperation.cs:
```diff
@@ -7,31 +7,118 @@
+            string or int or long or float or double or decimal or bool => obj,
+
+            // Handle Bytes
+            byte[] bytes => new Dictionary<string, object>
```
Should this be separated out/reused

YBTopaz8 replied:
I would see why!
But did not find the functional need to reorganize.
If commenting out works for now, I can do that

dblythy commented on Parse/Infrastructure/Control/ParseAddUniqueOperation.cs:
```diff
@@ -7,61 +7,125 @@
+            // Handle Bytes
+            byte[] bytes => new Dictionary<string, object>
+            {
+                ["__type"] = "Bytes",
```
Should this be separated out/reused

YBTopaz8 replied:
Would appreciate what would be better meant by "Should this be separated out/reused".
Do I not handle them explicitly? move to a new file?  where?
Do I would default to removing them from he file first?
Thanks.

dblythy replied:
No worries, my suggestion is whether the `EncodeForParse` method should be moved to a class + new file. If this logic is duplicative elsewhere, we can centralize it

YBTopaz8 replied:
Ah, I see now.
Well I would not have much of a say as I believe I followed the existing structure so if the method is there it means I had to include it perhaps to make it complete/work.

Personally, I'd suggest keeping, as It is ONLY used by this file.

YBTopaz8 replied:
We could find a why to centralize them though!

But I would let anyone else with the skills and time to help us with that (unless I find the time once again).

I would kindly remind that I focused more on bringing Parity and functionality + security.
I believed and still believe that code clean up could be well done with someone having THAT in mind specifically and clear directives.

dblythy replied:
No worries, happy to leave as is for now if the logic is not duplicative

dblythy commented:
Had a quick look and a fair few of the changes seem to be linting related - my suggestion would be to revert the File Scoped Namespaces change so it's a bit easier to review, and then address the File namespaces in a sepeate PR (that way this PR can just be the required changes)

Would also suggest modularising some of the LiveQuery code into its own files (the diff here has no added files which I think is strange)

If you could, could you let me know which classes have been added/changed in this PR (and isn't a lint change)

YBTopaz8 commented on Parse/Platform/Files/FileState.cs:
```diff
@@ -1,30 +1,48 @@
+    /// </summary>
+    private static bool IsParseHostedFile(Uri location)
+    {
+        return location.Host.EndsWith("parsetfss.com", StringComparison.OrdinalIgnoreCase);
```
I haven't found this version of the file in on my end.
Checking again

dblythy replied:
A fair bit of the diff seems to be converting to file scoped namespaces - the diff says 183 files/15k lines changed but I don't think there was actually that much edited, right?

YBTopaz8 replied:
Correct!

I did this change Mid port as a reflex since I usually use file scoped and by the time I noticed, I was too far gone to revert my changes, so I moved on as such.
Reason why I felt better when UT passed, At least they ensured the code is functionally upright👍🏾

dblythy replied:
No worries. Just a bit hard to review all the files 😃 

Could you please let me know which files introduce the new functionality / have been actually changed and i'll focus my review there

YBTopaz8 replied:
Sure thing.
I'll do my best to point them out.
let me push this commit and brb

YBTopaz8 commented:
Okay, it might be a very long reply;
File Changed : Parse/Infrastructure/HostManifestData.cs
- Updated `public static HostManifestData Inferred` for MAUI compatibility (should work for Unity as well)

File: Parse/Utilities/ObjectServiceExtensions.cs
- Updated GetFieldForPropertyName as previous impl didn't account for some cases

Files: Parse/Infrastructure/Data/ParseDataDecoder.cs & Parse/Infrastructure/Data/ParseObjectCoder.cs
- Did a overhaul, pretty much

File: Parse/Infrastructure/Execution/ParseCommandRunner.cs
- Fixed a bug where then new way of receiving errors from Parse wasn't done, so positive error messages came out as Exceptions (simply had to receive them as docs)
- Updated the way exceptions are handled. It's more graceful and crashes only when dev needs to act. 
In fact the whole lib is as such now - It should not crash if it can simply return null or so.

File: Parse/Platform/Objects/MutableObjectState.cs
- Updated to handle more data-better

Files : Parse/Platform/Objects/ParseObjectClass.cs and Parse/Platform/Objects/ParseObjectClassController.cs

File: Parse/Utilities/UserServiceExtensions.cs
- Updated methods to new async/await ways.

File: Parse/Infrastructure/Utilities/JsonUtilities.cs 
- Updated for Parity (better encoding/returns)

Files: Parse/Platform/Users/ParseUserController.cs , Parse/Platform/Users/ParseCurrentUserController.cs,  Parse/Platform/Users/ParseUser.cs
- Overhauled 
File: Parse/Platform/Installations/ParseInstallationController.cs
- Fixed some issues with installation ID making it crash on some plats (will still comeback for those Mutex if need be).

File : Parse/Abstractions/Platform/Objects/IObjectState.cs 
- Added INotifyPropertyChanged follow a UT, I suppose it's useful!

Parse/Abstractions/Platform/Objects/IParseObjectController.cs,
- SaveAll and DeleteAll now expect Any Enumerable and not just ILists. Users can now use arrays etc..

Parse/Infrastructure/Utilities/Conversion.cs
- Now accounts for most if not all file types

Parse/Platform/Security/ParseACL.cs
- Updated to handle nested ACLs too if any

Added Support for .net 5,6,7,8,9 and MAUI.
I didn't mention changed I made before (other async changes etc)

I updated this `<PackageReference Include="OpenCover" Version="4.7.1221" />`

Here is a full overview of what's done @dblythy  !

Also, about Parse Live Query, I added No file because I believe it's best to keep it as an optional library.
I do have one all setup and updated (I pushed an update yesterday to now use RX.NET to allow for LINQ support - I believe .NET will be the only platform with LINQ support on LQ which is quite cool!

If the Parse Community is interested, I can help make it officially its.
Just let me know! I will _Not_ be handling any Unit Tests though - sorry! Doing that for Parse was enough field day for me.

The Live Queries Lib is online and already having little over 70 downloads in Nugets , no Negative Feedback since it's out and even internally, I've had no issues.

If interested, of course, kindly ping me.

dblythy commented:
Thank you for providing that documentation @YBTopaz8! No worries, I misunderstood - was thinking the livequery code was part of this PR. Having it as a seperate package is clean. I'll go in again tomorrow and review the classes you mentioned.

Appreciate your contribution - this will be a valuable uplift!

YBTopaz8 commented:
It's a pleasure!
I find the potential for Parse and especially Live Queries (and now with LINQ) to make it much much better for any kind of users (especially the power users who do a lot of data manipulation).

I'll Actually try to do a full documentation of what's changed what's added/removed (probably nothing if not bugs) and how to use new methods I added.

It's a pleasure to help again, and I appreciate you all!
Lmk If there's any changes done that need a review!

mtrezza commented:
I also assumed LiveQuery was part of this PR. So we need to remove that from the PR title, right?

YBTopaz8 commented:
Right - on a technically!

It DOES fully support LQ compared to before, but doesn't Ship With it.

The confusion may be understandable indeed.

On Wed, Dec 11, 2024, 8:12 a.m. Manuel ***@***.***> wrote:

> I also assumed LiveQuery was part of this PR. So we need to remove that
> from the PR title, right?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/parse-community/Parse-SDK-dotNET/pull/393#issuecomment-2535954452>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJ5TYCFJQTVHSW7WSNK7Q532FA24ZAVCNFSM6AAAAABTBRJWT2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDKMZVHE2TINBVGI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

YBTopaz8 commented:
If this block is caused by code, do let me know

When I review my code again for CacheController, the logic doesn't seem fully right.
I'm highly suspecting the new locks implemented as they are in the trace.
This got my attention because setting a LQ in my v1.0 release can't work with ACL so I'll dive back on those.
If LQ fully works then the base Parse should be Flawless and Fully presentable.

Cheers.

YBTopaz8 commented:
Updating to inform that it looks we can't have LiveQueries with ACL as parameters.
So I'll just Parse throws the right error and polish up the lib again.

YBTopaz8 commented:
Both Fixed.
Committing soon.

mtrezza commented:
We seems to have a low coverage in this PR. According to the CI this PR reduces the coverage from 53% to <1%. Can that be true?

YBTopaz8 commented:
May I know what is known by "Coverage" in this context?
Would they be the Tests?

mtrezza commented:
We use codecov as a tool that measures how much code is covered by tests. It basically checks which code lines are executed during the test run `ci / check-ci` to calculate the coverage. See https://github.com/parse-community/Parse-SDK-dotNET/pull/393#issuecomment-2539729762.

mtrezza commented:
I see in the logs that the tests seem to be running for several .NET versions, is that correct?

```
Test run for D:\a\Parse-SDK-dotNET\Parse-SDK-dotNET\Parse.Tests\bin\Debug\net6.0\Parse.Tests.dll (.NETCoreApp,Version=v6.0)
Test run for D:\a\Parse-SDK-dotNET\Parse-SDK-dotNET\Parse.Tests\bin\Debug\net7.0\Parse.Tests.dll (.NETCoreApp,Version=v7.0)
Test run for D:\a\Parse-SDK-dotNET\Parse-SDK-dotNET\Parse.Tests\bin\Debug\net8.0\Parse.Tests.dll (.NETCoreApp,Version=v8.0)
Test run for D:\a\Parse-SDK-dotNET\Parse-SDK-dotNET\Parse.Tests\bin\Debug\net9.0\Parse.Tests.dll (.NETCoreApp,Version=v9.0)
```

If so, maybe it makes sense to run the tests separately for each .NET version, instead of running them all in a single CI job.

YBTopaz8 commented:
Well, for all I know, I added compatibility for .net5,6,7,8 and 9 and when I run tests, the UT runs for all of them too.

I can put for .net 9 only then.

I'm pretty new to tests so I'll see how to have them log their names.

Expect a commit.

YBTopaz8 commented:
Well, for all I know, I added compatibility for .net5,6,7,8 and 9 and when I run tests, the UT runs for all of them too.

I can put for .net 9 only then.

I'm pretty new to tests so I'll see how to have them log their names.

Expect a commit.

mtrezza commented:
Could you set this up so that the tests run with the .NET version that is set in the env var `DOTNET_VERSION`? Then I'll change the CI to run the job with each version.

YBTopaz8 commented:
Okay - so before I commit (as I'm all set now)
I can set the .NET version to .NET 8 now while you update the `DOTNET_VERSION` on your end then when good, you just let me know and I'll push the commit.

What says you?

mtrezza commented:
I was referring to the supported versions. Supported version simply means we run the unit tests in that .NET environment and they pass. It doesn't mean that the Parse SDK code needs to be converted to 8.x syntax.

I've changed the CI to run the tests in different .NET environments. I'm actually not sure why the CI was running was already running in [4 different .NET versions](https://github.com/parse-community/Parse-SDK-dotNET/pull/393#issuecomment-2540273676). I'd assume that only the CI sets up the .NET environment with the specific versions (which it does now) and the unit tests just run in that environment, without caring about the .NET version.

YBTopaz8 commented:
Alright.
I think I get it all.
I'll commit both Parse and Parse.Test now, and they both have .NET 8 as their target versions, inline with the CI

YBTopaz8 commented:
This should be the last commit from me, I believe.
The only UT failure I expect is LogOutAsync and so because I couldn't wrap my head around creating a mock that will account for device storage manipulation (after logging out) so the test will fail despite it doing through.

mtrezza commented:
It seems that all the tests are running in .NET 8.0. If you take a look at the logs, for example for .NET 6.x:

```
Executing: C:\Program Files\dotnet\dotnet.exe
  Determining projects to restore...
  All projects are up-to-date for restore.
  Parse -> D:\a\Parse-SDK-dotNET\Parse-SDK-dotNET\Parse\bin\Debug\net8.0\Parse.dll
  Parse.Tests -> D:\a\Parse-SDK-dotNET\Parse-SDK-dotNET\Parse.Tests\bin\Debug\net8.0\Parse.Tests.dll
Test run for D:\a\Parse-SDK-dotNET\Parse-SDK-dotNET\Parse.Tests\bin\Debug\net8.0\Parse.Tests.dll (.NETCoreApp,Version=v8.0)
VSTest version 17.12.0 (x64)
```

Where in code do you define the .NET version that the tests should run in?

mtrezza commented:
Should we really change to multiple target frameworks? Wouldn't that create multiple builds? I think if we target only .NET 6.0, then we'd ensure that it is compatible with .NET 7.0, 8.0, 9.0, because they are backwards compatible, right? And in a separate PR we'd change to target .NET 8.0 as that's a breaking change.

YBTopaz8 commented:
If ci is set to test on 6,7,8,9 then we are ok!
here is the cproj
```

    <PropertyGroup>
        <TargetFrameworks>net6.0;net7.0;net8.0;net9.0;</TargetFrameworks> <!--Here are the target frameworks!-->
        <DocumentationFile>bin\Release\netstandard2.0\Parse.xml</DocumentationFile>
        <Version>3.0.2</Version>
        <LangVersion>latest</LangVersion>

        <Authors>Parse</Authors>
        <PackageProjectUrl>https://parseplatform.org/</PackageProjectUrl>
        <RepositoryUrl>https://github.com/parse-community/Parse-SDK-dotNET/</RepositoryUrl>
        <PackageIconUrl></PackageIconUrl>
        <RepositoryType>GitHub</RepositoryType>
        <Description>This is the official package for the Parse .NET Standard SDK. Add a cloud backend to any platform supporting .NET Standard 2.0.</Description>
        <Copyright>Copyright © Parse Platform</Copyright>
```

YBTopaz8 commented:
I agree with you 100% on this one @mtrezza 
I don't know why it's more hectic that I thought and the PR is indeed too long
I'll Send a (hopefully) final commit targeting ONLY .NET 6 in both the tests and Parse itself.
Parse Live Queries (and now v2) will both work on .NET 6, no worries on that when time comes,

If this is fine by you, please cancel the ci, and I'll push the commit.

mtrezza commented:
Let's see if I get this right, there are 2 stages in the CI:

- Building: There's a single .NET version in which we build the library, that can be .NET 9.0 for example. That is where the `<TargetFramework>` is important. If we set multiple `<TargetFramework>`, then that will result in multiple builds, one for each .NET version. But we're only uploading a single build to Nuget, so we won't actually need multiple builds, so I assume we only need a single `<TargetFramework>`.

- Testing: Then we want to test the build in multiple .NET versions. The CI installs multiple .NET versions in the environment by default, so we'd probably need to tell it in which version we want to run the tests, or uninstall the other .NET versions.

YBTopaz8 commented:
I felt like it more of an explainer for me, as it's highly valuable!

And that is how it works, then as we discussed, kindly target ONLY .NET6 for now, based on our very latest discussion

YBTopaz8 commented:
That means we set one target and preferably one for both parse & test.

I set it to .net 6

Underwhelming a little bit the PR, but that's alright I suppose.

mtrezza commented:
I'm trying to get the CI to use the proper .NET for testing...

mtrezza commented:
The CI should now run the tests using each respective .NET version. The trick was to set the `<TargetFrameworks>net6.0;net7.0;net8.0;net9.0</TargetFrameworks>` only for the test project `Parse.Tests.csproj`.

To recap:
- We are compiling the Parse .NET SDK with `<TargetFramework>net6.0</TargetFramework>` which implies compatibility with future versions of .NET.
- When adding or removing a .NET version we need not just to modify `ci.yml` for the CI but also the `<TargetFrameworks>` in `Parse.Tests.csproj`.

YBTopaz8 commented:
Crossing fingers!

mtrezza commented:
This is generally not a high coverage to begin with. I think we could merge this if someone could tests the SDK out a bit, to ensure it works fine. Then we can gradually improve coverage over time.

YBTopaz8 commented:
Yes!

I am using it in my project - building a cross plat, cross sync music player app 
https://github.com/YBTopaz8/Dimmer-MAUI

And my sample Maui chat app that uses the latest version of LQ since it's the demo project.
https://github.com/YBTopaz8/Chat-App-.NET-MAUI-LIVE-QUERY

I'm fine with waiting as long as it fits the guidelines as well

mtrezza commented:
@YBTopaz8 Before we merge, I'd like to confirm with you that this PR doesn't contain any breaking changes, is that correct? If so, we'll release this as a feature release, otherwise we'd release as a major release.

YBTopaz8 commented:
Correct, this PR does not contain any breaking changes as far as I've checked. It's more of a "facelift," modernizing the codebase and readying it for Live Query integration, so it's good to go.

To ensure reliability, I’ve extensively tested it across platforms, including Android WearOS. It worked flawlessly, even in a real-time chat app with full CRUD functionality running across multiple platforms: [Here](https://github.com/YBTopaz8/Chat-App-.NET-MAUI-LIVE-QUERY). I also built a native watch app for my cross-platform music app ([Here too](https://github.com/YBTopaz8/DimmerWatchCompanion)).

If I had a smart fridge, I’d have tested it there too! 😄
That said, I believe this qualifies as a major release because it modernizes the codebase and officially drops support for older .NET versions. Of course, I'm happy to wait if further review or consolidation is required.

YBTopaz8 commented:
Do Grant me a moment though.
I would like to push a final commit removing unnecessary debug lines (that print out both dumb and valuable data)
Should be done in a bit

mtrezza commented:
> officially drops support for older .NET versions.

Does it? That would quality as a breaking change. I thought the breaking change would come with the next PR where we'll remove the .NET 6 and 7 support?

YBTopaz8 commented:
Ohh, you would be right indeed!
I suppose .NET 6 has support for previous versions too so - Yes I was mistaken!
No breaking change there.

mtrezza commented:
> I suppose .NET 6 has support for previous versions too

I don't think there is a guaranteed backwards compatibility, we're targeting .NET 6 in this PR, so it should work with .NET >=6, but not necessarily <6. Which .NET version does the SDK target so far, before this PR? If it's also .NET 6 then there's no change.

mtrezza commented:
Did some research. Currently the Parse .NET SDK is targeting `netstandard2.0` which is old but provides compatibility with many platforms like Unity, Xamarin, etc. That will be gone with this PR which targets .NET.

I think this change makes strategic because that's the future:
- Microsoft stopped development of netstandard, and is encouraging developers to shift to .NET.
- Unit 7 is working on migrating to [CoreCLR](https://discussions.unity.com/t/unity-future-net-development-status/836646/2403), but that will take time, maybe with Unity 7 in 2025 / 2026.
- Xamarin has evolved into .NET MAUI and Microsoft has stopped Xamarin support on May 1, 2024.

So yes, this is a breaking change. Given that the Parse .NET SDK didn't receive much TLC for years I'd say let's go ahead with this modernization.

YBTopaz8 commented:
This is right !
Well, In the case where apps target .net (which is understandable tbh!) can use v.3.0.2
Those who can switch to .net 6 targeting will simply update to latest.

Since Xamarin has evolved into .NET MAUI and is now unsupported (as of May 1, 2024), the decision to drop netstandard2.0 eliminates reliance on outdated technologies.

& Also 
Unity's move to CoreCLR is still a work in progress and won’t be fully ready until Unity 7 (likely 2025/2026). For now, Unity developers targeting .NET Standard can continue using older versions of the Parse SDK if needed.

And I'm pretty sure I fixes to MAUI does fix unity in most if not all cases as they seem to operate pretty much same.

& Finally, I fixed the logout issue found in unity only that moved to Maui since they are mostly treated the same as MAUI. 
Issue was: The client instance used to logout had too much "unnecessary" info that would, for some reasons lead to a 403 - unauthorized request instead of a 200-OK or 400-Bad.
so after numerous back and forth with postman, I could confirm that.

I'll add a simple condition to use a vanilla webclient with only the data needed for very simple operations.

Expect a final commit fixing that . (Maybe I'll fix the unit test too as well! Finally after months!)

YBTopaz8 commented:
Just to reassure you, pray least seeing this another around would me
1. Leaving support for old tech - it's a logical choice now even for companies since new .net versions have clear differences in performance they'd naturally enjoy + security robustness. 
2. It solves many unity issues that carried to .net and thank I fixed likely out of necessity but didn't know since I might not best suited for that! (But if I have it so working on a Watch then it failing on unity would be interesting to see!)

3. These code changes are fully compatible with the Parse Live Queries SDK - if that was worry of yours - no need!

All the changes do here are tested WITH LQ first, then with UTests.

I'm wondering how Parse will manage that for .net

Will you make it a standalone extension ? Or merge like Parse for Swift Did? You tell me in due time of course!

YBTopaz8 commented:
Unit Tests all run as expected. Logout still fails because I can't mock it properly.
I give up for now at least.

I fixed my main issue which was that it would freeze/crash app.

And since we opted to release as breaking change, I went ahead and implemented the Todos that required methods.
Also exposed some others and error handling should be much better

![image](https://github.com/user-attachments/assets/8bedf8e1-b632-4591-9f40-eb0bec14886b)

YBTopaz8 commented:
Perfection.

parseplatformorg commented:
🎉 This change has been released in version [4.0.0](https://github.com/parse-community/Parse-SDK-dotNET/releases/tag/4.0.0)

mtrezza commented:
I suggest you post this comment on the https://github.com/parse-community/Parse-SDK-dotNET/issues/199 to have a discussion. I noticed that the issue was linked incorrectly, because it was about adding LiveQuery to the SDK, but this PR did not add LiveQuery, so I've reopened it.
